game/dino_jump_main.py

```diff
@@ -0,0 +1,394 @@
+        self.SPEED = speed  # number of pixels moved to the left per loop
+        self.ARRAY = import_sprite(
+            self.OBS_TYPE
+        )  # calls sprites: with type to get pixel array
+        self.x = init_width  # initial position of spirt at right edge of board
+        self.HEIGHT, self.WIDTH = (
+            self.ARRAY.shape
+        )  # height and width of the array to call
+
+        # Inherited from Game
+        self.GROUND_AXIS, self.LOW_BIRD_AXIS, self.HIGH_BIRD_AXIS = AXES
+
+        # Define the y axis this obstacle sprite travels along
+        if self.OBS_TYPE in ("c", "C"):
+            self.y = self.GROUND_AXIS - self.HEIGHT
+        elif self.OBS_TYPE == "b":
+            self.y = self.LOW_BIRD_AXIS - self.HEIGHT
+        else:
+            self.y = self.HIGH_BIRD_AXIS - self.HEIGHT
+
+
+#############################################################################
+# OBSTACLE MANAGEMENT MODULE
+#############################################################################
+
+
+class ObstacleManager:
+
+    def __init__(self, WIDTH, AXES):
+
+        # Timing Constants
+        self.curr_cooldown = (
+            0  # counts the number of frames before next obstacle generated
+        )
+        self.BASE_COOLDOWN = (
+            20  # must wait how many frames before new obstacle generates
+        )
+        self.COOLDOWN_VARIANCE = np.linspace(
+            10, 30, 5
+        )  # variance of how many frames to add the to baseline
+
+        # Inherited from game
+        self.WIDTH = WIDTH
+        self.AXES = AXES
+
+        # Possible Choices for Compound Obstacles and Speeds
+        self.ALL_SPEEDS = [3]
+        self.ALL_OBSTACLE_GROUPS = ["c", "C", "ccc", "CC", "cCc", "cc", "b", "B"]
+
+        # Hold all current obstacle objects to keep track of
+        self.curr_obs = []
+
+    def update_location(self):
+
+        # Only care about valid obstacles on the game board
+        self.curr_obs = [obs for obs in self.curr_obs if obs.x + obs.WIDTH > 0]
+
+        # Update each obstacle's location by its speed
+        for obs in self.curr_obs:
+            obs.x -= obs.SPEED
+
+    def choose_random_obstacle_attributes(self):
+        # Choose the next obstacle: cactus arrangement, low bird, or high bird
+        return np.random.choice(self.ALL_OBSTACLE_GROUPS), np.random.choice(
+            self.ALL_SPEEDS
+        )
+
+    def generate_obstacle(self):
+
+        # When the cooldown reaches zero, a new obstacle is generated
+        if self.curr_cooldown <= 0:
+            new_obstacle, new_speed = self.choose_random_obstacle_attributes()
+            for i, obs in enumerate(new_obstacle):
+                new_obs = SingleObstacle(obs, new_speed, self.WIDTH, self.AXES)
+                new_obs.x += (new_obs.WIDTH + 1) * i
+                self.curr_obs.append(new_obs)
+            self.curr_cooldown = self.BASE_COOLDOWN + random.choice(
+                self.COOLDOWN_VARIANCE
+            )

```

Comment:
I might just be bad at the game lol, but have we done any analysis on the base_cooldown + cooldown variance and ensuring it's always possible to not die?

Reply:
We haven't yet, we can probably link all the variables through some calculation though when initialising but currently we can adjust each as sees fit, which I think will be helpful for testing especially when we first start integrating with camera, in case there are some latencies

Reply:
The current set of variables should be very possible to not die though! It's a tight jump on the triple cactis :)